packages/devtools_app/lib/src/standalone_ui/api/impl/dart_tooling_api.dart

```diff
@@ -82,43 +86,153 @@ class DartToolingApiImpl implements DartToolingApi {
+class PostMessageEditorClient implements EditorClient {
+  PostMessageEditorClient(this._api) {
+    // In PostMessage world, we just get events with the entire new list so
+    // we must figure out what the actual changes are so we can produce the
+    // same kinds of events as the new DTD version.
+    _api.devicesChanged.listen((e) {
+      final supportedDevices = e.devices.map(
+        (d) => EditorDevice.fromJson({...d.toJson(), 'supported': true}),
+      );
+      final unsupportedDevices = e.unsupportedDevices?.map(
+            (d) => EditorDevice.fromJson({...d.toJson(), 'supported': false}),
+          ) ??
+          [];
+      final newDevices =
+          supportedDevices.followedBy(unsupportedDevices).toList();
+      final newIds = newDevices.map((d) => d.id).toSet();
+      final oldIds = _currentDevices.map((d) => d.id).toSet();
+
+      // Devices that are not in the new set have been removed.
+      for (final id in oldIds.difference(newIds)) {
+        _eventController.add(DeviceRemovedEvent(deviceId: id));
+      }
+      // Devices in the new set have either been changed or were added.
+      for (final device in newDevices) {
+        if (oldIds.contains(device.id)) {
+          _eventController.add(DeviceChangedEvent(device: device));
+        } else {
+          _eventController.add(DeviceAddedEvent(device: device));
+        }
+      }
+      // And record the updated set.
+      _currentDevices
+        ..clear()
+        ..addAll(newDevices);
+
+      // Finally, handle if the selection changed.
+      if (e.selectedDeviceId != _currentSelectedDeviceId) {
+        _currentSelectedDeviceId = e.selectedDeviceId;
+        _eventController.add(
+          DeviceSelectedEvent(deviceId: _currentSelectedDeviceId),
+        );
+      }
+    });
+    _api.debugSessionsChanged.listen((e) {
+      final newIds = e.sessions.map((d) => d.id).toSet();
+      final oldIds = _currentDebugSessions.map((d) => d.id).toSet();
+
+      // Sessions that are not in the new set have been removed.
+      for (final id in oldIds.difference(newIds)) {
+        _eventController.add(DebugSessionStoppedEvent(debugSessionId: id));
+      }
+      // Sessions in the new set have either been changed or were added.
+      for (final session in e.sessions) {
+        if (oldIds.contains(session.id)) {
+          _eventController.add(DebugSessionChangedEvent(debugSession: session));
+        } else {
+          _eventController.add(DebugSessionStartedEvent(debugSession: session));
+        }
+      }
+      // And record the updated set.
+      _currentDebugSessions
+        ..clear()
+        ..addAll(e.sessions);
+    });
+
+    // Trigger the initial initialization now we have the handlers set up.
+    // In the old postMessage world, this is how we get the initial set of
+    // devices/sessions.
+    unawaited(_api.initialize());
   }
 
-  @protected
-  final json_rpc_2.Peer rpc;
+  final VsCodeApi _api;
```

Comment:
is this code expected to be IDE agnostic in the future? Will there be one IdeApi that both VS Code and IntelliJ confirm to, or will we maintain two separate APIs - one for VS code and one for IJ?

Reply:
It's not, this is the "legacy" API. In the complete PR that adds the DTD bits, this gets renamed to `PostMessageToolApi` (https://github.com/flutter/devtools/pull/7919/files#diff-10d85e415369fb2761370af4cc25dceab8b5f50bd4a432fdfd40bb6ee8310714). Some of that API is mapped onto the new DTD classes (`EditorDevice`, `EditorDebugSession`) in this PR, but the main interface of this is postMessage only and won't be used when communicating with DTD (the `EditorClient` class is essentially the replacement).